hub/core/transform/test_transform.py

```diff
@@ -0,0 +1,212 @@
+        ds_out = ds
+        ds_out.create_tensor("image")
+        ds_out.create_tensor("label")
+        fn2(copy=1, mul=2).eval(data_in, ds_out, num_workers=5)
+        assert len(ds_out) == 99
+        for index in range(1, 100):
+            np.testing.assert_array_equal(
+                ds_out[index - 1].image.numpy(), 2 * index * np.ones((index, index))
+            )
+            np.testing.assert_array_equal(
+                ds_out[index - 1].label.numpy(), 2 * index * np.ones((1,))
+            )
+
+        assert ds_out.image.shape_interval.lower == (99, 1, 1)
+        assert ds_out.image.shape_interval.upper == (99, 99, 99)
+
+
+@enabled_datasets
+def test_chain_transform_list_small(ds):
+    ls = [i for i in range(100)]
+    ds_out = ds
+    ds_out.create_tensor("image")
+    ds_out.create_tensor("label")
+    pipeline = hub.compose([fn1(mul=5, copy=2), fn2(mul=3, copy=3)])
+    pipeline.eval(ls, ds_out, num_workers=5)
+    assert len(ds_out) == 600
+    for i in range(100):
+        for index in range(6 * i, 6 * i + 6):
+            np.testing.assert_array_equal(
+                ds_out[index].image.numpy(), 15 * i * np.ones((337, 200))
+            )
+            np.testing.assert_array_equal(
+                ds_out[index].label.numpy(), 15 * i * np.ones((1,))
+            )
+
+
+@enabled_datasets
+@pytest.mark.xfail(raises=NotImplementedError, strict=True)
+def test_chain_transform_list_big(ds):
+    ls = [i for i in range(2)]
+    ds_out = ds
+    ds_out.create_tensor("image")
+    ds_out.create_tensor("label")
+    pipeline = hub.compose([fn3(mul=5, copy=2), fn2(mul=3, copy=3)])
+    pipeline.eval(ls, ds_out, num_workers=3)
+    assert len(ds_out) == 8
+    for i in range(2):
+        for index in range(4 * i, 4 * i + 4):
+            np.testing.assert_array_equal(
+                ds_out[index].image.numpy(), 15 * i * np.ones((1310, 2087))
+            )
+            np.testing.assert_array_equal(
+                ds_out[index].label.numpy(), 15 * i * np.ones((13,))
+            )
+
+
+@enabled_datasets
+def test_chain_transform_list_small_processed(ds):
+    ls = list(range(100))
+    ds_out = ds
+    ds_out.create_tensor("image")
+    ds_out.create_tensor("label")
+    if isinstance(remove_memory_cache(ds.storage), MemoryProvider):
+        with pytest.raises(InvalidOutputDatasetError):
+            fn2().eval(ls, ds_out, num_workers=3, scheduler="processed")
+        return
+
+    pipeline = hub.compose([fn1(mul=5, copy=2), fn2(mul=3, copy=3)])
+    pipeline.eval(ls, ds_out, num_workers=3, scheduler="processed")
+    assert len(ds_out) == 600
+    for i in range(100):
+        for index in range(6 * i, 6 * i + 6):
+            np.testing.assert_array_equal(
+                ds_out[index].image.numpy(), 15 * i * np.ones((337, 200))
+            )
+            np.testing.assert_array_equal(
+                ds_out[index].label.numpy(), 15 * i * np.ones((1,))
+            )
```

Comment:
can you add 2 more tests:
- check that threaded is faster than serial
- check that processed is faster than threaded

this is pretty important because hub 1.x had the problem that processed was like 5x slower than threaded at one point

Reply:
the tests can be very simple but it would just be nice to have, especially when running these tests on multiple OSs

Reply:
processed won't always be faster. For computationally expensive tasks of course, but for I/O tasks threaded may run faster.
For these tests do you propose timing 2 transforms in the same test, noting the time and asserting that t1<t2?